c/json.c

```diff
@@ -1760,10 +2154,81 @@ int jsonAsBoolean(Json *json) {
 int jsonAsNumber(Json *json) {
   if (json->type == JSON_TYPE_NUMBER) {
     return json->data.number;
+  } else if (json->type == JSON_TYPE_INT64){
+    return json->data.integerValue;
   }
   return 0;
 }
 
+int64_t jsonAsInt64(Json *json){
+  switch (json->type){
+  case JSON_TYPE_NUMBER:
+    return (int64)json->data.number;
+  case JSON_TYPE_INT64:
+    return json->data.integerValue;
+  case JSON_TYPE_DOUBLE:
+    return (int64)json->data.floatValue;
+  default:
+    return 0;
+  }
+}
+
+double jsonAsDouble(Json *json){
+  switch (json->type){
+  case JSON_TYPE_NUMBER:
+    return (double)json->data.number;
+  case JSON_TYPE_INT64:
+    return (double)json->data.integerValue;
+  case JSON_TYPE_DOUBLE:
+    return json->data.floatValue;
+  default:
+    return 0;
+  }
+}
+
+static int64 hashString(char *s){
+  int64 hash = 5381;
+  int i;
+  int len = strlen(s);
+
+  for (i=0; i<len; i++){
+    hash = ((hash << 5) + hash) + s[i];
+  }
+  return hash&0x7fffffffffffffffll;
+}
+
+int64_t jsonLongHash(Json *json){
+  int64_t hash = ((int64_t)json->type) << 56;
```

Comment:
Why do we need the type value in the 8 highest bits? Does it improve value distribution?

Reply:
Yes.  Also, this function is more experimental than sane.  I don't like it and hope to remove it.  Eventually we need a tree-walking jsonEquals method.